### PR TITLE
[Merged by Bors] - feat: port Analysis.Calculus.MeanValue

### PR DESCRIPTION
**Pull request by urkud**
---

- [x] depends on: #4508

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#4221
* leanprover-community/mathlib4#4239
* leanprover-community/mathlib4#4241
* leanprover-community/mathlib4#4242
* leanprover-community/mathlib4#4247
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#4221~~
* ~~leanprover-community/mathlib4#4239~~
* ~~leanprover-community/mathlib4#4241~~
* ~~leanprover-community/mathlib4#4242~~
* ~~leanprover-community/mathlib4#4247~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by fpvandoorn** [comment]
(I assume this was `awaiting-review`? Otherwise, sorry)

File: Mathlib/Analysis/Calculus/MeanValue.lean [hunk 1]
```diff
@@ -0,0 +1,1365 @@
...
+variable {𝕜 G : Type _} [IsROrC 𝕜] [NormedSpace 𝕜 E] [NormedAddCommGroup G] [NormedSpace 𝕜 G]
+
+namespace Convex
+
+variable {f g : E → G} {C : ℝ} {s : Set E} {x y : E} {f' g' : E → E →L[𝕜] G} {φ : E →L[𝕜] G}
+
+open AffineMap in
+/-- The mean value theorem on a convex set: if the derivative of a function is bounded by `C`, then
+the function is `C`-Lipschitz. Version with `has_fderiv_within`. -/
+theorem norm_image_sub_le_of_norm_has_fderiv_within_le
```
Comment (fpvandoorn) [inline Mathlib/Analysis/Calculus/MeanValue.lean:461]:
```suggestion
theorem norm_image_sub_le_of_norm_hasFderivWithin_le
```
Various lemma names still have to be fixed, I was a bit too hasty with merging

Comment (Parcly-Taxel) [reply Mathlib/Analysis/Calculus/MeanValue.lean]: Should be done now.

**Review by fpvandoorn — APPROVED** [review]
I'll give @urkud another chance to look over the changes, but LGTM

**Comment by urkud** [comment]
bors merge

**Pull request closed without merging**